### PR TITLE
Updated to empirical protein models (which previously only contained …

### PR DESCRIPTION
**Pull request by sjspielman**
…WAG) to include these additional empirical models: JTT, LG, JC.

### CODE REVIEW

File: res/TemplateBatchFiles/libv3/models/protein/empirical.bf [hunk 1]
```diff
@@ -306,6 +322,779 @@ models.protein.WAG.empirical_Q = {
   }
 };
 
+
+
+//==================================================================================================================//
+//===================================== SJS added LG, JC, and JTT below =================================================//
```
Comment (stevenweaver) [inline res/TemplateBatchFiles/libv3/models/protein/empirical.bf]: git blame gives credit where credit is due. :)

**Comment by sjspielman** [comment]
Comments pushed. Should be good to merge.

**Comment by spond** [comment]
Dear @sjspielman,

Great work! Thanks for adding those. One small request: I have been meaning (in older versions of HyPhy as well) to "normalize" the rates to so that the expected number of substitutions per site is 1 when model-specific frequencies are used. In other words, if we have q<sub>ij</sub> (rate matrix) and &pi;<sub>j</sub> (frequencies), then 

<code>
&sum;<sub>i &neq j</sub> q<sub>_ij &pi; <sub>j</sub> &pi; <sub>i</sub> = 1
<code>

Best,
Sergei

**Comment by sjspielman** [comment]
Hi @spond, 

Can definitely do this - are you looking for a function to normalize rates (which may necessitate some further discussion about libv3 idiosyncrasies which aren't immediately clear to me how to navigate), or simply a redefined matrix? It's clear to me how to make a new exchangeability matrix, but not (yet) a new Q matrix in the context of libv3 organization/philosophy.

Thanks, 
Stephanie

**Comment by spond** [comment]
Dear @sjspielman,

Either approach would work. I think having a normalization function would be a more "bug-proof" approach, but we need to make it optional, in case people desire to use different scaling. 

At the moment, the easiest thing to do is to specify the exchangeability matrix in scaled rates (conditioned on model frequencies).

Best,
Sergei

**Comment by sjspielman** [comment]
## To be run using commit 582b7ecc3898b403920af5a43eced3393b7fb68a

See attached example-error.zip for relevant files.

```HYPHYMP LIBPATH=`pwd`/res `pwd`/res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/ProteinGTRFit.bf```
\
[example-error.zip](https://github.com/veg/hyphy/files/799919/example-error.zip)
# Provide /path/to/filelist.txt and enter 1 for resume cached analysis.

Clearly, branch length optimization is fine if you are starting an analysis from scratch. However, if you load a cache file, then branch length optimization fails with some alphabet issue.
Datasets appear to be read in in the same manner for REV fits and BL optimizations, however, so unclear why this step is causing the problem.


Full output from such a run:
```
Analysis Description
--------------------
Fit a general time reversible model to a collection of training protein
sequence alignments. Generate substitution and scoring matrices
following the procedures described in Nickle et al 2007

- __Citation__: Nickle DC, Heath L, Jensen MA, Gilbert PB, Mullins JI, Kosakovsky Pond
SL (2007) HIV-Specific Probabilistic Models of Protein Evolution. PLoS
ONE 2(6): e503. doi:10.1371/journal.pone.0000503

- __Written by__: Sergei L Kosakovsky Pond and Stephanie J Spielman

- __Contact Information__: {spond,stephanie.spielman}@temple.edu

- __Analysis Version__: 0.01


Supply a list of files to include in the analysis (one per line) (`/Users/sjspielman/Research/HyPhy-repos/hyphy/res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/`) /Users/sjspielman/temp/filelist.txt


			+----------------------------------------------------------------------------------------------+
			|A cache file of a prior analysis on this list of files was detected. Would you like to use it?|
			+----------------------------------------------------------------------------------------------+


	(1):[YES] Resume analysis using the detected cache file.
	(2):[NO] Launch a new analysis and *overwrite* the detected cache file.

 Please choose an option (or press q to cancel selection):1


### Initial branch length fit


[PHASE 1] Performing initial branch length optimization using 0

### Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = 0
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = 0
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = 0
Overall Log(L) = -20337.90105046801


[PHASE 2] Performing initial REV fit to the data

### Loaded cached results for 'REV-Phase-0'. Log(L) = -19808.37375119318


[PHASE 3] Iteratively optimizing branch lengths and fitting REV model until convergence.

### Retuning branch lengths (Rev-local-Phase-1)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6634.495266910425
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4276.194329857094
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8884.898240379809
Overall Log(L) = -19795.58783714733

### Loaded cached results for 'REV-Phase-1'. Log(L) = -19800.95248348704

### Retuning branch lengths (Rev-local-Phase-2)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6636.547912290107
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4275.666615333657
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8886.119642916799
Overall Log(L) = -19798.33417054056

### Loaded cached results for 'REV-Phase-2'. Log(L) = -19808.56600453644

### Retuning branch lengths (Rev-local-Phase-3)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6639.906352144378
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4275.904412732011
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8887.688171051806
Overall Log(L) = -19803.4989359282

### Loaded cached results for 'REV-Phase-3'. Log(L) = -19816.56537318978

### Retuning branch lengths (Rev-local-Phase-4)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6639.100143030688
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4274.334273113262
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8895.846528883612
Overall Log(L) = -19809.28094502756

### Loaded cached results for 'REV-Phase-4'. Log(L) = -19793.84608311106

### Retuning branch lengths (Rev-local-Phase-5)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6635.423349915791
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4275.011141100412
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8882.297297436038
Overall Log(L) = -19792.73178845224

### Loaded cached results for 'REV-Phase-5'. Log(L) = -19762.70267726184

### Retuning branch lengths (Rev-local-Phase-6)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6633.286389485734
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4263.501211637793
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8863.196700984077
Overall Log(L) = -19759.9843021076

### Loaded cached results for 'REV-Phase-6'. Log(L) = -19798.46725297734

### Retuning branch lengths (Rev-local-Phase-7)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6641.823152184053
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4268.75685949758
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8885.498773408752
Overall Log(L) = -19796.07878509038

### Loaded cached results for 'REV-Phase-7'. Log(L) = -19777.50654259434

### Retuning branch lengths (Rev-local-Phase-8)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6637.824291832617
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4265.146781992592
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8872.152397698694
Overall Log(L) = -19775.1234715239

### Loaded cached results for 'REV-Phase-8'. Log(L) = -19752.53922355745

### Retuning branch lengths (Rev-local-Phase-9)
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3. Log(L) = -6636.175373922196
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot2.dat' 2/3. Log(L) = -4258.435008701495
Loaded cached results for '/Users/sjspielman/temp/protgtr_fitter_alignments/prot3.dat' 3/3. Log(L) = -8857.145624810662
Overall Log(L) = -19751.75600743435

### Fitting the REV model using constrained branch lengths proportions (REV-Phase-9)
Current Max: -19761.134     (90 % done) LF Evals/Sec: 105.4   CPU Load: 2.694   
### Retuning branch lengths (Rev-local-Phase-10)
Dispatching file '/Users/sjspielman/temp/protgtr_fitter_alignments/prot1.dat' 1/3
Error:
The declared model alphabet 'null' does not match the jYxazbDU.nuc_data_0 filter: '{
{"A", "C", "D", "E", "F", "G", "H", "I", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "V", "W", "Y"}
}' in call to assert(

Function call stack
1 : Assert 'Columns(alphabet)==Columns(givenAlphabet)&&model.MatchAlphabets(givenAlphabet,alphabet)'
-------
2 : models.generic.AttachFilter(xVlDYwZa.model,_filter_)
-------
3 : ExecuteCommands in string transform using basepath .
-------
4 : Assert 'Columns(alphabet)==Columns(givenAlphabet)&&model.MatchAlphabets(givenAlphabet,alphabet)'
-------
5 : models.generic.AttachFilter(model.generic.DefineModel.model,data_filter)
-------
6 : user_model=model.generic.DefineModel(model_template,name_space,{"0":"terms.global"},filters,None)
-------
7 : protein_gtr.rev_file_mle=estimators.FitSingleModel_Ext(utility.Map(protein_gtr.filter_specification,"_value_","_value_['name']"),utility.Map(protein_gtr.partitions_and_trees,"_value_","_value_['tree']"),protein_gtr.rev_model_branch_lengths,protein_gtr.rev_file_mle,None)
-------
8 : Call(result_callback,0,Eval(job+'('+Join(",",utility.Map(arguments,"_value_","utility.convertToArgumentString (_value_)"))+')'),arguments)
-------
9 : mpi.QueueJob(protein_gtr.queue,"protein_gtr.fitGTRtoFile",{"0":protein_gtr.file_list[file_index],"1":protein_gtr.current_gtr_fit["global"],"2":(protein_gtr.current_gtr_fit[terms.json.attribute.branch_length])[(protein_gtr.current_gtr_fit["filename-to-index"])[protein_gtr.file_list[file_index]]]},"protein_gtr.handle_gtr_callback")
-------
10 : protein_gtr.phase_results=protein_gtr.run_gtr_iteration_branch_lengths()
-------

Check errors.log for details on execution errors.

Check messages.log details of this run.
```

File: res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/ProteinGTRFit.bf [hunk 2]
```diff
@@ -21,6 +21,7 @@ LoadFunctionLibrary("ProteinGTRFit_helper.ibf"); // Moved all functions from thi
 /*------------------------------------------------------------------------------*/
 
 utility.ToggleEnvVariable ("NORMALIZE_SEQUENCE_NAMES", 1);
+utility.SetEnvVariable ("OPTIMIZATION_PRECISION", 0.1); // for testing purposes.
```
Comment (stevenweaver) [inline res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/ProteinGTRFit.bf]: Are we sure we want to leave this here?

Comment (sjspielman) [inline res/TemplateBatchFiles/SelectionAnalyses/ProtGTR_Fitter/ProteinGTRFit.bf]: We are not sure! We are further not sure about this merge. I may have (definitely did?) do something backwards on my end, wherein I meant to merge some changes from Sergei into my fork before pushing to veg, but this happened in reverse order.

**Pull request merged**